bitswap.go

```diff
@@ -307,18 +315,38 @@ func (bs *Bitswap) receiveBlocksFrom(from peer.ID, blks []blocks.Block) error {
 	// to the same node. We should address this soon, but i'm not going to do
 	// it now as it requires more thought and isnt causing immediate problems.
 
-	// Send all blocks (including duplicates) to any sessions that want them.
+	allKs := make([]cid.Cid, 0, len(blks))
+	for _, b := range blks {
+		allKs = append(allKs, b.Cid())
+	}
+
+	wantedKs := allKs
+	if len(blks) != len(wanted) {
+		wantedKs = make([]cid.Cid, 0, len(wanted))
+		for _, b := range wanted {
+			wantedKs = append(wantedKs, b.Cid())
+		}
+	}
+
+	// Send all block keys (including duplicates) to any sessions that want them.
 	// (The duplicates are needed by sessions for accounting purposes)
-	bs.sm.ReceiveBlocksFrom(from, blks)
+	bs.sm.ReceiveBlocksFrom(from, allKs)
 
-	// Send wanted blocks to decision engine
-	bs.engine.AddBlocks(wanted)
+	// Send wanted block keys to decision engine
+	bs.engine.AddBlocks(wantedKs)
```

Comment:
You mean to determine whether a block is unique or a duplicate? We do already call `Has()` for accounting purposes before we get to this point so it might simplify things to just use that  to determine unique/dup

Reply:
(I'd suggest we do that in a follow-up PR)

Reply:
Wait, nevermind. I was referring to Bitswap.Has but Bitswap. Has calls _this_.